crates/ruff_linter/src/rules/pylint/rules/invalid_string_characters.rs

```diff
@@ -201,8 +219,97 @@ pub(crate) fn invalid_string_characters(
         let location = range.start() + TextSize::try_from(column).unwrap();
         let range = TextRange::at(location, c.text_len());
 
-        diagnostics.push(Diagnostic::new(rule, range).with_fix(Fix::safe_edit(
-            Edit::range_replacement(replacement.to_string(), range),
-        )));
+        new_diagnostics.push(InvalidCharacterDiagnostic {
+            diagnostic: Diagnostic::new(rule, range),
+            // This is integrated with other fixes and attached to the diagnostic below.
+            edit: Edit::range_replacement(replacement.to_string(), range),
+        });
```

Comment:
We also need to handle the case where the invalid character was the last character before the quotes in a triple quoted strings:

```python
"""test"<invalid>"""
```

Let's say `<invalid>` is the invalid character. Removing `<invalid>` then results in `"""test""""` which is not a valid non-raw strings.